// stylelint-browser-compat/lib/types/color/__space_separated_parameters/collect_features.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.collectFeatures = collectFeatures;
const RULE_NAME_MAP = {
  hsl: `types.color.hsl.space_separated_parameters`,
  hsla: `types.color.hsl.space_separated_parameters`,
  rgb: `types.color.rgb.space_separated_parameters`,
  rgba: `types.color.rgb.space_separated_parameters`
};
const FUNCTION_NAME_LIST = ['hsl', 'hsla', 'rgb', 'rgba'];
function isTargetFunctionNameList(value) {
  return FUNCTION_NAME_LIST.includes(value);
}
async function collectFeatures({
  node,
  parent
}) {
  const funcName = node.value.replace(/^-(?:ms|o|webkit|moz)-/, '');
  if (!isTargetFunctionNameList(funcName)) {
    return [];
  }
  const features = [];
  const offset = parent.prop.length + (parent.raws.between?.length ?? 0);
  if (node.nodes.length === 5 && node.nodes[1]?.type === 'space' && node.nodes[3]?.type === 'space' || node.nodes.length === 7 && node.nodes[1]?.type === 'space' && node.nodes[3]?.type === 'space' && node.nodes[5]?.type === 'div' && node.nodes[5].value === '/') {
    const index = offset + node.sourceIndex;
    const endIndex = offset + node.sourceEndIndex;
    features.push({
      endIndex,
      id: RULE_NAME_MAP[funcName],
      index,
      name: `Space-separated ${funcName}() parameters`,
      node: parent
    });
  }
  return features;
}
