// stylelint-browser-compat/lib/at-rules/counter-style/collect_features.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.collectFeatures = collectFeatures;
var _descriptor_name_set = require("./descriptor_name_set");
async function collectFeatures({
  node
}) {
  const features = [];
  node.walkDecls(delc => {
    const [vendorPrefix] = /^-(webkit|moz|ms|o)-/.exec(delc.prop) ?? [];
    if (_descriptor_name_set.descriptorNameSet.has(delc.prop)) {
      features.push({
        endIndex: delc.prop.length,
        id: `at-rules.counter-style.${delc.prop}`,
        index: 0,
        name: `"${delc.prop}" descriptor of the @counter-style`,
        node: delc
      });
    } else if (vendorPrefix != null) {
      const name = delc.prop.replace(vendorPrefix, '');
      if (_descriptor_name_set.descriptorNameSet.has(name)) {
        features.push({
          endIndex: delc.prop.length,
          id: `at-rules.counter-style.${name}`,
          index: 0,
          name: `"${name}" descriptor of the @counter-style`,
          node: delc,
          prefix: vendorPrefix
        });
      }
    }
  });
  return features;
}
