// stylelint-browser-compat/lib/at-rules/page/collect_features.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.collectFeatures = collectFeatures;
var _postcssValueParser = _interopRequireDefault(require("postcss-value-parser"));
var _descriptor_name_set = require("./descriptor_name_set");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
async function collectFeatures({
  ignoreFeatures,
  node
}) {
  const promises = [];
  node.walkDecls(delc => {
    const [vendorPrefix] = /^-(webkit|moz|ms|o)-/.exec(delc.prop) ?? [];
    if (_descriptor_name_set.descriptorNameSet.has(delc.prop)) {
      promises.push(Promise.resolve([{
        endIndex: delc.prop.length,
        id: `at-rules.page.${delc.prop}`,
        index: 0,
        name: `"${delc.prop}" descriptor of the @page`,
        node: delc
      }]));
    } else if (vendorPrefix != null) {
      const name = delc.prop.replace(vendorPrefix, '');
      if (_descriptor_name_set.descriptorNameSet.has(name)) {
        promises.push(Promise.resolve([{
          endIndex: delc.prop.length,
          id: `at-rules.page.${name}`,
          index: 0,
          name: `"${name}" descriptor of the @page`,
          node: delc,
          prefix: vendorPrefix
        }]));
      }
    }
    const valueRoot = (0, _postcssValueParser.default)(delc.value);
    if (delc.prop === 'size') {
      promises.push(Promise.resolve().then(() => _interopRequireWildcard(require("./size/collect_features"))).then(({
        collectFeatures
      }) => {
        return collectFeatures({
          nodes: valueRoot.nodes,
          parent: delc
        });
      }));
    }
    promises.push(Promise.resolve().then(() => _interopRequireWildcard(require("../../properties/collect_features"))).then(({
      collectFeatures
    }) => {
      return collectFeatures({
        ignoreFeatures,
        node: delc,
        valueRoot
      });
    }));
    promises.push(Promise.resolve().then(() => _interopRequireWildcard(require("../../types/collect_features"))).then(({
      collectFeatures
    }) => {
      return collectFeatures({
        nodes: valueRoot.nodes,
        parent: delc
      });
    }));
  });
  const features = (await Promise.all(promises)).flat();
  return features;
}
